lib/ambient_context.ml

```diff
@@ -27,7 +27,7 @@ let without_binding k cb =
    let (module Store : STORAGE) = Atomic_.get storage in
    Store.without_binding k cb
 
-
+(* Why does this need to be synchronized? *)
 let with_storage_provider new_storage cb : unit =
```

Comment:
Not that it's bad to be safe, but if I understand the use case correctly, this is the top-level entry point function to install a single storage backend? Not sure why `cas` synchronization is necessary here?

Reply:
So, it's not documented yet, and currently racey, too, until I properly implement copying — but you're _supposed_ to be able to _change_ the storage at runtime.

Rather the point of this whole thing is to have a 'single source of truth' in your binary, that disparate libraries can agree on. The problem is that that source of truth *has* to change: at boot-up, and in some libraries persistently, there is only TLS. But once an asynchronous event-loop starts, TLS becomes unreliable.

(Similarly, although I haven't looked into it in detail yet, there is of course the case of multiple event-loops — I need to look into how `Eio_lwt` actually _plugs_ the Lwt event-loop; I suspect we can just use Eio's storage in the case of both, but I'm not sure? But in any case, it's likely that in a partially-converted Lwt/Eio program, additional “source-of-truth” moves may become necessary at runtime.)

Reply:
Ah, realized I didn't answer the original question —

well, currently, the configured storage is _not_ thread-local; it's a shared ref, common to all threads. The CAS was necessary to make it safely mutable … but, now that I'm thinking about it, maybe the configured-storage _should_ be, itself, thread-local? Hrm.

Reply:
> but you're supposed to be able to change the storage at runtime.

Not sure I follow here. My understanding of how this all works is that the storage backend is configured _once_ at runtime prior to the start of the event loop to select the appropriate Eio/Lwt storage provider. So as a consequence, the code is single threaded during the inital mutation.

> The problem is that that source of truth has to change: at boot-up, and in some libraries persistently, there is only TLS. But once an asynchronous event-loop starts, TLS becomes unreliable.

I lose you here about the "source of truth changes in some libraries persistently". What's an example of this?  I was under the impression that the changing "storage backend" is just a wrapper to configure the appropriate concurrency primatives the project uses. Then the management of "thread-local" context is handled by the underlying `Lwt.with_value` function

Reply:
> the storage backend is configured once at runtime prior to the start of the event loop to select the appropriate Eio/Lwt storage provider. So as a consequence, the code is single threaded during the inital mutation.

Even if this were true, the event-loop is completely separate from threads — still need to be thread-safe (or in this case, thread-specific), before the event-loop has booted up! It's trivially possible for a program to have forked _before_ it gets to whatever part involves an event-loop (and thus, whatever part might be configuring `ambient-context`.)

> Then the management of "thread-local" context is handled by the underlying Lwt.with_value function

Unfortunately, no — there's three _completely separate_ storage mechanisms; and they have to be converted between at runtime, in edge-cases where "program boot", "some dependency's library initialization", "systhread-fork", and "greenthread event-loop entry" don't happen in the order you want them to:

> What's an example of this?

1. program boots up
2. libraries are initialized, think `let () = Mylib.init …` running
    * a library calls something like `let context_key = Ambient_context.create_key ()`
    * that key now maps to the _thread-local storage_ (a particular and concrete store that is used by default when there's no event-loop)
3. the top-level code reaches a systhread fork
    * the key exists in both threads' separate TLS
4. the top-level code boots up the event loop, and calls `Ambient_context.with_storage_provider …` to configure it
    * oh-oh, now `Mylib.context_key` exists in TLS, _which won't be accessed anymore_
    * when the top-level app — or `Mylib`, even — tries to retrieve it, because now the active storage is, say, LWT, it gets … nothing

tl;dr "we can't synchronize when dependencies and apps attempt to create/access keys; and we can't guarantee it's _after_ a `with_storage_provider` call; even if, in an ideal world, there's only _one_ `with_storage_provider` call."